text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+
+    /// Attempt to resolve the future to a final value, registering
+    /// the current task for wakeup if the value is not yet available.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    /// - `Poll::Pending` if the future is not ready yet.
+    /// - `Poll::Ready(val)` with the result `val` of this future if it completed.
+    ///
+    /// Once a future has completed, clients should not `poll` it again.
+    ///
+    /// When a future is not ready yet, `poll` returns `Poll::Pending`.
+    /// The future will *also* register the interest of the current task in the
+    /// value being produced. For example, if the future represents the availability
+    /// of data on a socket, then the task is recorded so that when data arrives,
+    /// it is woken up (via `cx.waker()`). Once a task has been woken up,
+    /// it should attempt to `poll` the future again, which may or may not
+    /// produce a final value at that time.
+    ///
+    /// Note that if `Pending` is returned it only means that the *current* task
+    /// (represented by the argument `cx`) will receive a notification. Tasks
+    /// from previous calls to `poll` will *not* receive notifications.
+    ///
+    /// # Runtime characteristics
+    ///
+    /// Futures alone are *inert*; they must be *actively* `poll`ed to make
+    /// progress, meaning that each time the current task is woken up, it should
+    /// actively re-`poll` pending futures that it still has an interest in.
+    /// Usually this is handled automatically by `async`/`await` notation or
+    /// via adapter methods. Executors ensure that each task is `poll`ed every
+    /// time a future internal to that task is ready to make progress.
+    ///
+    /// The `poll` function is not called repeatedly in a tight loop for
+    /// futures, but only whenever the future itself is ready, as signaled via
+    /// `cx.waker()`. If you're familiar with the `poll(2)` or `select(2)`
+    /// syscalls on Unix it's worth noting that futures typically do *not*
+    /// suffer the same problems of "all wakeups must poll all events"; they
+    /// are more like `epoll(4)`.
+    ///
+    /// An implementation of `poll` should strive to return quickly, and should
+    /// *never* block. Returning quickly prevents unnecessarily clogging up
+    /// threads or event loops. If it is known ahead of time that a call to
+    /// `poll` may end up taking awhile, the work should be offloaded to a
+    /// thread pool (or something similar) to ensure that `poll` can return
+    /// quickly.
+    ///
+    /// # Panics
+    ///
+    /// Once a future has completed (returned `Ready` from `poll`), subsequent
+    /// calls to `poll` may panic, block forever, or otherwise cause bad behavior.
+    /// The `Future` trait itself provides no guarantees about the behavior of
+    /// `poll` after a future has completed.
+    fn poll(self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Item>;
```

Comment:
Changing this to `self: Pin<Self>` for the core future is a big change, of which I cannot anticipate the implications.

I commented [here](https://github.com/rust-lang/rfcs/pull/2395#issuecomment-379473677) with some questions about what this would mean.

Could the implications of having `self: Pin<Self>` on the ecosystem be discussed?

* How much additional unsafe code is required by users?
* How does `select` work now?
* How does this change impact fns that use `poll_ ... -> Poll<..>` but do not impl `Future`?
* How does this impact other traits like `Stream`, `Sink`, `AsyncRead`, `AsyncWrite`, `tower::Service`, etc...

Reply:
> But also most libraries should use async or future combinators rather than implementing Future themselves.

That probably isn't the case. At least, not for several years. Reasons why you wouldn't:

- You can't return `impl Future` from a trait method yet.
- You cannot name the the type, so it can't be wrapped in another type without being generic.
- You may wish to include other trait impls or inherent methods on the return type. Like in `Connection::graceful_shutdown()`, which can be called in between polls.
- You may wish to compose various `poll_*` methods, without converting each into a separate future, for performance.

It seems possible these issues will all be fixed in a couple years, but it'd be *really* painful to write libraries before then.

Reply:
@bill-myers 

> But also most libraries should use async or future combinators rather than implementing Future themselves.

On what are you basing this? Have you looked at the h2 library and understood what it is doing? My *entire* point is that nobody actually knows if what you are saying is true because nobody has tried it.

On top of all the reasons that @seanmonstar pointed out, it simply is not possible because the way h2 implements futures isn't one simple thread of logic, but there are many independent future impls that go through a coordinated state machine (potentially across multiple threads). This **definitely** cannot be implemented using future combinators and AFAIK it cannot be implemented with `async` / `await` either.

Reply:
> As for select, I guess it needs to be changed to take both futures inside a PinBox.

So, what you are suggesting is a common primitive that is currently zero cost will now incur allocation overhead? If so, this should be called out in the RFC in a drawback section or something, because it's not a minor detail.

Reply:
It is definitely not necessary to use `PinBox` for `select` as long as you don’t care about getting the unresolved future back when one completes. If you do then you will need a version of `select` that takes `Future + Unpin`, which can be satisfied by either `PinBox<Future>`, or `Pin<Future>` procured via some future stack pinning API.

Reply:
If you are implementing `Future` manually for some type then you can just not `impl !Unpin for Foo` and use the `impl<T: Unpin> DerefMut for T` to convert the `Pin<Self>` to `&mut Self`. The main complication is if you take in generic futures you will have to take them as `Future + Unpin` and call them via `Pin::new(&mut foo).poll(ctx)`.

Reply:
Also, there should be *zero* unsafe code required for implementing non-self-referential futures. I think we’re still missing one or two small APIs to allow implementing maybe-self-referential adaptors nicely without unsafe, but I think 99% of the operations needed can be hidden behind a safe API provided on `Pin`.

Reply:
## `Future::poll(Pin<Self>)` is leaky.

I still think exposing `self: Pin<Self>` in the `Future` trait is a leaky abstraction.

I understand the problem that `Pin` is trying to solve, and it's fantastic the thought put behind it so that we might have easy `async` and `await` in Rust. **However**, I'm personally left with the feeling that this proposal took this approach: "This allows performant async functions, which is easy for users, and the downsides for implementors don't matter as much as gaining love from new users". I don't want us to stop there. I don't want such a trade off forever frozen in libcore. I definitely don't want large 3rd party libraries eventually making their own `Future`s instead (Scala comes to mind).

Here's what I find leaky about this. Fundamentally, **a `Future` is a type that will return a value at some later point.** That's **all** the trait should be about. Being pinned is, frankly, unrelated to returning a future result.

The only time I'd care about a pinned future is when all of these are true:

- I want to use `async fn` (and not implement `Future` for a type).
- I want to borrow bindings across `await!` calls.

If I'm not doing those, then I don't need to be pinned. Yet, the current proposal requires that *if I ever want to implement `Future`*, I must always think about pinning.

## Pinning futures restricts safe futures

As mentioned in other comments, if **all** futures must be pinned, that makes other things much harder (even if they didn't actually need the pinning).

- `Select` doesn't have a proposed solution. (For those not familiar, `Select` polls 2 different futures, completing when either one finishes. It returns the first result, and the other future, so you can optionally schedule to poll it some more, or toss it. This requires moving the second future after it has been polled).
- Things like `poll_ready` would be unusable if they needed to poll some internal future. `poll_ready` would need to be changed as well to take a `Pin<Self>`, which means that the `Service` is then immovable.
- In my experience, I have at times needed to poll a future when I first get it, as an optimization, to see if the value is available already, or if I need to then spawn it on an `Executor`. I could instead always spawn onto the `Executor`, but now my libraries are taking a forced performance hit, to allow for pinned futures, even if the future didn't need to be pinned.


## `Iterator::next` does not take `Pin<Self>`

Consider the `Iterator` trait, and how it currently takes `&mut self`. There is the goal of eventually allowing generators to be used as iterators. How is that going to work? Why can't we take **that** solution, and use it for `Future` as well?

```rust
// where fibonacci is a generator-iterator
for i in fibonacci().map(|x| x * 5).take(10) {
    // ...
}
```

I don't imagine that `Iterator::next` will be changed, as that's a breaking change. It also seems unlikely that the iterator adapter would box the generator. And it'd be an unacceptable requirement to force the user to pin the generator before using it (and impossible if you're already generic over `Iterator`).

Some possibilities that we could explore more:

- Some combination of sub traits and specialization. `Future::poll(&mut self)` and `Async::poll_pinned(Pin<Self>)`, with blanket impls such that pinning an `Async` makes it implement `Future`. There'd still need to be something to thread the `Pin` down... Though, since `async fn` is already mutating the actual function, the `Pin` could be threaded through like that.
- Perhaps knowledge of being pinned could be in `Context`. Needing to be pinned is already extremely unsafe, it seems like handling that in a `Future` for a self-referential generator could just check `if cx.pinned()` or similar.

Reply:
@Nemo157 

> Also, there should be zero unsafe code required for implementing non-self-referential futures. I think we’re still missing one or two small APIs to allow implementing maybe-self-referential adaptors nicely without unsafe, but I think 99% of the operations needed can be hidden behind a safe API provided on Pin.

Then all of this information should be part of the RFC.

Right now, as proposed, getting `&mut self`  from `self: Pin<Self>` requires [unsafe code](https://github.com/rust-lang/rust/blob/8228d8e17649bab2881ac77c31e32ba6e341fa64/src/libcore/mem.rs#L1156).

This means that, as proposed, in order to port the existing ecosystem to what is proposed will require adding extensive amounts of `unsafe` blocks to what is currently safe code.

Reply:
@Nemo157 

> It is definitely not necessary to use PinBox for select as long as you don’t care about getting the unresolved future back when one completes. 

This is something that I *do* care about and have used extensively, not just with `select`. A common pattern I use is to have a type that impls `Future` representing the completion of a complex computation.  Then, I use other futures that take this complex future and yields it back to represent steps in the complex computation.

A simple example would be a future that represent a connection. This future completes when the connection closes, then, that type has `fn ready(self) -> Ready<Self>` that takes the connection and completes yielding the connection back once the connection is ready to dispatch a request.

Now, **as far as I know**, based on the RFC, this pattern becomes either a) not possible or b) requires unsafe code.

What I ask is for some additional information related to how all these patterns need to be updated. Saying "well, you just shouldn't do it" isn't reasonable because all these patterns are driven by real needs.

You mention a whole bunch of things with `!Unpin` and `Deref`, etc.. that I'm not following. If these are ways to deal with the limitations imposed by the RFC, then they should be outlined in the RFC.

Reply:
> Right now, as proposed, getting `&mut self` from `self: Pin<Self>` requires unsafe code.

[Unless the type is `Unpin`](https://github.com/rust-lang/rust/blob/8228d8e17649bab2881ac77c31e32ba6e341fa64/src/libcore/mem.rs#L1187), as long as you are not explicitly marking your type as `Unpin`, or including an `Unpin` member, then you can use this `DerefMut` implementation.

[Here's a quick sketch of a manual `Future` implementation using the new trait](https://play.rust-lang.org/?gist=e182c3d8d607e193f8964b0c71031b24&version=nightly), note that there is no `unsafe` code used, and mutating the future object inside `poll` transparently uses the `DerefMut` to avoid any need to use the `Pin` APIs.

Reply:
And, [here are examples of the two ways of implementing `select`](https://play.rust-lang.org/?gist=c3942a616dbfbe26cf0f42607ba12d4b&version=nightly).

The first (`Select`) is the only real possibility for immovable `Future`s since you can't move them to return them, it uses a single bit of `unsafe` code to go from its pin to a pin of one of its fields. I believe this is guaranteed to be safe for a normal field access, so something like the `pin_field!` macro is one of the potential missing APIs that could be provided by `std`; which would make this example 100% safe code. If you look at the first and second blocks in `main` for polling this one you can see that it is a maybe-immovable adaptor, if it gets passed `Unpin` futures then it will itself be `Unpin`, but if it gets passed an immovable future then it will be immovable.

The second (`SelectUnpin`) is the current `select` trimmed down to remove the `Error` (and written from scratch, so no idea how close the internals are).  It only operates on movable `Future`s, but if you look at the fourth block in `main` you can see that it's possible to pass an immovable `Future` in by pinning it yourself and using a forwarding `impl<F: Future> Future for PinBox<F>` (and the same for `Pin` if we had stack pinning).

Reply:
I'm a little confused.. Isn't an important aspect of `select` that it returns the futures that have not yet completed alongside the result of the one that did? I don't see how you do that with pinned futures?

Reply:
Seems @seanmonstar beat me to it in https://github.com/rust-lang/rfcs/pull/2395#issuecomment-383268534.

Reply:
@Nemo157 One thing worth noting: the `pin_field!` macro is actually *not* safe for arbitrary !Unpin types (which I think is pretty unacceptable since, as you can see, it makes it close to impossible to use pinned futures without using unsafe).  However, I may have come up with a way to mitigate that (essentially, allowing a !Unpin type to opt into proving that it can be pinned).  It's hard to do in pure library code, even using procedural macros, for various reasons (in particular: if specialization lands, I'm not yet sure how to enforce a key invariant, that `Drop` must not be directly implemented for `!Unpin` types that support safe field projections), but I think it can be done.  In your particular case, the projection is definitely safe.